### PR TITLE
Introduce prepared application templates (#641)

### LINKED ISSUE #641: Introduce PreparedApp and one-time route/middleware preparation
## Summary

Introduce an immutable prepared-application representation that consumes `WireframeApp`'s mutable registration state and builds route middleware chains exactly once.

This is the foundational implementation slice for ADR 012 proposal #637. It should establish the internal boundary without yet changing the server's public factory evaluation semantics; the following server-runtime issue will switch connection tasks onto the prepared root.

## Problem

`WireframeApp` currently stores both:

- builder inputs: `handlers`, `middleware`, callback registrations, protocol/assembler definitions, and configuration;
- runtime cache: `OnceCell<Arc<HashMap<u32, HandlerService<E>>>>`.

`handle_connection_result` lazily initializes the route table and clones its inner `Arc`. When the server constructs a fresh app per connection, the cache has no opportunity to amortize its work across connections.

The route map and completed middleware chains are immutable runtime data. They should be produced by an explicit preparation transition rather than hidden inside connection handling.

## Proposed design

Add an internal or initially crate-private transition shaped approximately as:

```rust
impl<S, C, E, F> WireframeApp<S, C, E, F> {
    pub(crate) async fn prepare(self) -> Result<PreparedApp<S, C, E, F>, PrepareError>;
}
```

`PreparedApp` should own:

- `HashMap<u32, HandlerService<E>>` directly;
- serializer and codec template/configuration;
- application data;
- lifecycle callback definitions;
- protocol and message assembler definitions;
- push, fragmentation, timeout, and memory-budget configuration.

Preparation should consume `handlers` and `middleware`, constructing each middleware chain once. Avoid retaining both the original handler map and a cloned route-service map after preparation unless a documented runtime need requires both.

### Required

- Define `PreparedApp` and any focused `PrepareError` type.
- Move `build_chains` out of the per-connection path.
- Consume or otherwise finalize route and middleware registrations during preparation.
- Store the prepared route table by value, not as `OnceCell<Arc<HashMap<...>>>`.
- Provide an internal compatibility path so existing tests can prepare an app before driving a connection.
- Keep the builder's type-changing methods working before preparation.

### Deferred

- Evaluating `AppFactory` once during server startup.
- Public constructor/deprecation decisions for `WireframeServer`.
- Full `ConnectionRuntime` extraction.
- Protocol ownership consolidation.
- Replacing every callback/assembler `Arc`; that should happen after the prepared root is in use.

## Acceptance criteria

- [ ] `WireframeApp` has an explicit consuming preparation transition.
- [ ] `PreparedApp` contains a direct prepared route table.
- [ ] Route middleware transforms run exactly once per preparation, proven with counters.
- [ ] Two or more connections driven from one `PreparedApp` reuse the same route services without rebuilding chains.
- [ ] Duplicate route detection remains a builder-time error.
- [ ] Route ordering and middleware ordering remain unchanged.
- [ ] Preparation failure is typed and does not partially expose a runtime object.
- [ ] No `OnceCell<Arc<HashMap<...>>>` remains in the prepared runtime representation.
- [ ] Direct app-driving tests and testkit helpers have a clear migration path.
- [ ] #598 is reviewed: accessors made obsolete by the prepared representation are removed or explicitly retained and tested.
- [ ] #639 connection-startup instrumentation records factory/transform counts before and after.

## Tests

- Prepare an app with multiple routes and middleware layers; assert transform counters equal one per layer/route as intended.
- Drive multiple in-process connections from one prepared app; assert no additional transform calls.
- Assert route dispatch and middleware ordering match current tests.
- Assert a failed transform/preparation returns an error before any runtime is available, if middleware transformation can fail after the supporting error work lands.
- Add a compile-time or type-level test that route registration is unavailable after transitioning to `PreparedApp`.

## Design constraints

- Do not make `PreparedApp` mutable behind a lock.
- Do not clone the entire builder to preserve registrations after preparation.
- Do not introduce a second public route-registration surface on `PreparedApp`.
- Keep ADR 010's packet/transport-frame boundary unchanged.

## Dependencies

- Decision: ADR 012 proposal #637.
- Evidence: baseline issue #639.

This issue should land before the server-runtime and connection-runtime slices.

## References

- Epic #635
- ADR 011 proposal #636
- ADR 012 proposal #637
- #598
- #639
- `src/app/builder/core.rs`
- `src/app/builder/routing.rs`
- `src/app/inbound_handler.rs`
- `src/middleware.rs`

### PR DESCRIPTION
**Pull request by leynos**
## Summary

This branch introduces an immutable `PreparedApp` that consumes a
`WireframeApp` builder and transforms each route middleware chain once.
Prepared connections borrow the direct route table, so subsequent connections
do not rebuild route services.

Closes #641.

The server deliberately retains its existing per-connection factory evaluation
semantics. Deprecated builder-driving wrappers preserve the current direct-test
path until the server-runtime slice adopts the prepared root.

## Review walkthrough

- Start with [PreparedApp](https://github.com/leynos/wireframe/blob/issue-641-introduce-preparedapp-and-one-time-route-middleware-preparation/src/app/prepared_app.rs#L29) for the immutable runtime representation, consuming transition, retained protocol accessors, and compile-checked public API examples.
- Then review the [connection boundary](https://github.com/leynos/wireframe/blob/issue-641-introduce-preparedapp-and-one-time-route-middleware-preparation/src/app/inbound_handler.rs#L48) and [stream core](https://github.com/leynos/wireframe/blob/issue-641-introduce-preparedapp-and-one-time-route-middleware-preparation/src/app/inbound_handler/core.rs#L111), which borrow prepared routes without cloning or rebuilding them.
- Finish with the [startup counter harness](https://github.com/leynos/wireframe/blob/issue-641-introduce-preparedapp-and-one-time-route-middleware-preparation/tests/prepared_app.rs#L220), [users' guide migration](https://github.com/leynos/wireframe/blob/issue-641-introduce-preparedapp-and-one-time-route-middleware-preparation/docs/users-guide.md#L179), [compile-time boundary](https://github.com/leynos/wireframe/blob/issue-641-introduce-preparedapp-and-one-time-route-middleware-preparation/tests/ui/prepared_app_rejects_route.rs#L7), and [test-helper migration path](https://github.com/leynos/wireframe/blob/issue-641-introduce-preparedapp-and-one-time-route-middleware-preparation/wireframe_testing/src/helpers/drive.rs#L223).

## Validation

- `make check-fmt`: passed
- `make lint`: passed
- `make typecheck`: passed
- `make test`: passed
- `make markdownlint`: passed
- `make nixie`: passed
- `cargo test --doc`: passed
- `coderabbit review --agent`: completed with zero findings

## Notes

The startup harness records the #639 baseline with two routes and two
middleware layers: two legacy TCP connections invoke the factory twice and
perform eight transforms; one preparation adds a single factory invocation and
four transforms; two prepared connections add neither.

`PrepareError` is typed for future fallible middleware transforms. The current
transition is infallible, so it cannot expose a partial prepared runtime.

## References

- Closes [#641](https://github.com/leynos/wireframe/issues/641)
- [Runtime ownership baseline](https://github.com/leynos/wireframe/issues/639)
- [ADR 012 proposal](https://github.com/leynos/wireframe/issues/637)
- [Lody session](https://lody.ai/leynos/sessions/613f0dab-1989-4d5b-92c7-2c2122790af3)

## Summary by Sourcery

Introduce an immutable prepared application runtime that separates route registration from reusable connection handling while preserving legacy builder compatibility.

New Features:
- Add an immutable `PreparedApp` transition that builds route middleware chains once and supports reuse across connections.
- Provide prepared connection drivers and testing helpers for driving prepared applications.
- Expose preparation and prepared-connection observability metrics.

Bug Fixes:
- Preserve connection teardown execution when prepared stream processing returns an error.
- Propagate server I/O failures through testing drivers instead of swallowing them.

Enhancements:
- Retain deprecated builder-driven connection APIs and server factory semantics as compatibility paths while the runtime migration continues.
- Retain protocol and message-assembler accessors on prepared applications.
- Add compile-time and integration coverage for prepared application ownership, middleware reuse, concurrent connections, TCP handling, and middleware ordering.

CI:
- Align the main coverage workflow with the Wireframe repository and CodeScene project used by pull-request coverage checks.
- Add workflow contract tests to protect the CodeScene coverage baseline configuration.

Documentation:
- Document the prepared application lifecycle, compatibility behavior, testing migration path, and v0.3.0-to-v0.4.0 migration guidance.
- Update the roadmap and developer documentation to track prepared application adoption and follow-up server-runtime work.

Tests:
- Add deterministic instrumentation tests and integration coverage verifying one-time route preparation and reuse across connections.
- Add observability assertions for preparation duration, outcomes, and prepared connection usage.
- Add compile-fail coverage preventing route registration on prepared applications.

### CODE REVIEW

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

Do not treat warnings as optional or aspirational. Where a change is out of scope for this PR, propose a GitHub issue unless one exists already. (Treat o11y, code safety, documentation and validation coverage as in scope).

### Assessment against linked issues

| Issue | Objective | Addressed | Explanation |
| --- | --- | --- | --- |
| [#641](https://github.com/leynos/wireframe/issues/641) | Introduce an immutable PreparedApp with a consuming WireframeApp::prepare transition that owns the prepared route table and the application’s runtime configuration, without retaining the builder’s mutable registration state. | ✅ |  |
| [#641](https://github.com/leynos/wireframe/issues/641) | Build each route’s middleware chain during preparation and reuse those services across multiple connections, preserving route and middleware ordering and preventing middleware transforms from running again on the prepared connection path. | ✅ |  |
| [#641](https://github.com/leynos/wireframe/issues/641) | Provide the required migration and validation support, including typed preparation errors, a route-registration type boundary, prepared-app test helpers, and the requested factory/transform instrumentation evidence. | ❌ | The PR adds the typed PrepareError, compile-time boundary, prepared test helpers, and transform-reuse tests, but it does not add or update the [#639](https://github.com/leynos/wireframe/issues/639) connection-startup instrumentation to record factory and transform counts before and after. Additionally, the deprecated WireframeApp compatibility methods and the existing server connection spawner still build route chains per connection; only callers that explicitly prepare the app receive the one-time behavior. |

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

Do not treat warnings as optional or aspirational. Where a change is out of scope for this PR, propose a GitHub issue unless one exists already. (Treat o11y, code safety, documentation and validation coverage as in scope).

\[#641\](https://github.com/leynos/wireframe/issues/641) Provide the required migration and validation support, including typed preparation errors, a route-registration type boundary, prepared-app test helpers, and the requested factory/transform instrumentation evidence. ❌ The PR adds the typed PrepareError, compile-time boundary, prepared test helpers, and transform-reuse tests, but it does not add or update the \[#639\](https://github.com/leynos/wireframe/issues/639) connection-startup instrumentation to record factory and transform counts before and after. Additionally, the deprecated WireframeApp compatibility methods and the existing server connection spawner still build route chains per connection; only callers that explicitly prepare the app receive the one-time behavior.

File: wireframe_testing/src/helpers/drive.rs [hunk 1]
```diff
@@ -1,9 +1,14 @@
 //! Frame-oriented in-memory driving helpers.
 
+#![expect(
+    deprecated,
+    reason = "legacy test drivers preserve builder-based coverage during migration"
+)]
```
[the remark this answers is not in the document]
Comment (leynos) [reply wireframe_testing/src/helpers/drive.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: tests/ui/prepared_app_rejects_route.rs [hunk 2]
```diff
@@ -0,0 +1,15 @@
+use wireframe::{
+    app::{Envelope, Handler, WireframeApp},
+    serializer::BincodeSerializer,
+};
```
[the remark this answers is not in the document]
Comment (leynos) [reply tests/ui/prepared_app_rejects_route.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: src/app/prepared_app.rs [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+}
+
+impl<S, C, E, F> PreparedApp<S, C, E, F>
+where
+    S: Serializer + FrameMetadata<Frame = super::Envelope> + Send + Sync,
+    C: Send + 'static,
+    E: Packet,
+    F: FrameCodec,
+    super::Envelope: DecodeWith<S> + EncodeWith<S>,
+{
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/app/prepared_app.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: src/app/mod.rs [hunk 4]
```diff
@@ -23,10 +23,12 @@ mod memory_budgets;
...
+mod prepared_app;
 
 pub use builder::WireframeApp;
 pub use envelope::{Envelope, Packet, PacketParts};
-pub use error::{Result, SendError};
+pub use error::{PrepareError, Result, SendError};
 pub use lifecycle::{ConnectionSetup, ConnectionTeardown};
 pub use memory_budgets::{BudgetBytes, MemoryBudgets};
 pub use middleware_types::{Handler, Middleware};
+pub use prepared_app::PreparedApp;
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/app/mod.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: src/app/inbound_handler/core.rs [hunk 5]
```diff
@@ -0,0 +1,303 @@
...
+        message_assembler,
+        fragmentation,
+        memory_budgets,
+        read_timeout_ms,
+    } = context;
+    let codec = codec.clone();
+    let combined = CombinedCodec::new(codec.decoder(), codec.encoder());
+    let mut framed = Framed::new(stream, combined);
+    let requested_frame_length = codec.max_frame_length();
+    let max_frame_length = clamp_frame_length(requested_frame_length);
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/app/inbound_handler/core.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: src/app/inbound_handler.rs [hunk 6]
```diff
@@ -1,64 +1,128 @@
...
+        );
+        return Err(error);
+    }
+
+    if let (Some(teardown), Some(state)) = (on_disconnect, state) {
+        teardown(state).await;
+    }
+
+    Ok(())
+}
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/app/inbound_handler.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: src/app/prepared_app.rs [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+    /// Consume builder registrations and prepare immutable route services.
+    ///
+    /// Middleware transforms run once for every registered route during this
+    /// transition. The returned template can then drive multiple connections
+    /// without rebuilding its route chains.
+    ///
+    /// # Errors
+    ///
+    /// Returns [`PrepareError`] if a future fallible preparation step fails.
+    pub async fn prepare(self) -> Result<PreparedApp<S, C, E, F>, PrepareError> {
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/app/prepared_app.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: src/app/inbound_handler.rs [hunk 8]
```diff
@@ -69,257 +133,51 @@ where
...
-        }
-    }
-
-    /// Handle an accepted connection end-to-end, returning any processing error.
+    /// Handle a connection through a compatibility route preparation path.
     ///
     /// # Errors
     ///
     /// Returns an [`io::Error`] if stream processing or handler execution fails.
+    #[deprecated(note = "prepare the app once, then call PreparedApp::handle_connection_result")]
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/app/inbound_handler.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: wireframe_testing/src/helpers/drive.rs [hunk 9]
```diff
@@ -215,6 +220,47 @@ where
...
+        .map_err(|error| io::Error::other(error.to_string()))?;
+    drive_prepared_with_frames(&prepared, frames).await
+}
+
+/// Drive one connection through an already prepared application.
+///
+/// The borrowed prepared application can be driven repeatedly, allowing tests
+/// to verify that route middleware transforms are not rebuilt per connection.
+pub async fn drive_prepared_with_frames<S, C, E>(
+    app: &PreparedApp<S, C, E>,
```
[the remark this answers is not in the document]
Comment (leynos) [reply wireframe_testing/src/helpers/drive.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

**Comment by leynos** [comment]
@coderabbitai review

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

Do not treat warnings as optional or aspirational. Where a change is out of scope for this PR, propose a GitHub issue unless one exists already. (Treat o11y, code safety, documentation and validation coverage as in scope).

### ❌ Failed checks (2 errors, 8 warnings)

| Check name | Status | Explanation | Resolution |
| :---: | :--- | :--- | :--- |
| Module-Level Documentation | ❌ Error | Add module-level documentation to the new `tests/ui/prepared_app_rejects_route.rs` crate. The file starts with `use` and has no `//!` or `/*!` docstring. `tests/compile_error.rs` registers it with `tr… | Insert a leading docstring that states the fixture's purpose and relationship to the prepared-application API, for example `//! Compile-fail coverage showing that route registration is unavailable on PreparedApp.` Recheck all newly added an… |
| Rust Compiler Lint Integrity | ❌ Error | Fail the check because the PR adds broad module-level `#![expect(deprecated)]` suppressions in 21 testkit, test, and helper modules. The suppressions cover whole modules, including unrelated items, an… | Replace each module-level `#![expect(deprecated)]` with a narrow expectation on the specific legacy helper or call site, or narrow the module boundary so only compatibility code is compiled there. Add a code comment linking each unavoidable… |
| Docstring Coverage | ⚠️ Warning | Docstring coverage is 58.70% which is insufficient. The required threshold is 80.00%. Docstring coverage is scoped to functions touched by this diff. Analyzed 46 functions across 39 files. (1 skipped:… | Write docstrings for the functions missing them to satisfy the coverage threshold. |
| User-Facing Documentation | ⚠️ Warning | Document the new user-facing preparation API. The pull-request diff from merge base `cab4c95` to `2a60c7a` changes no documentation files. It adds the public `PreparedApp` type, `WireframeApp::prepare… | Update `docs/users-guide.md` with a user-facing section and working example that shows building a `WireframeApp`, calling `prepare().await`, handling connections through `PreparedApp`, reusing the prepared application, and handling `Prepare… |
| Developer Documentation | ⚠️ Warning | The pull request adds public `PreparedApp` and `PrepareError` APIs, changes the `WireframeApp` lifecycle, and adds prepared test-driver helpers. The diff contains no developer documentation changes. `… | Update `docs/developers-guide.md` to document the implemented builder-to-prepared lifecycle, including `WireframeApp::prepare().await`, its `Result<PreparedApp<...>, PrepareError>` outcome, one-time route middleware transformation, immutabl… |
| Testing (Unit And Behavioural) | ⚠️ Warning | Fail the testing check. The pull request adds a useful integration test for the successful `PreparedApp` path: it uses the public preparation and connection APIs, records factory and middleware-transf… | Add tests at the public boundary. Drive a prepared app with malformed input or a failing transport and assert the `io::Error` returned by `PreparedApp::handle_connection_result`; verify the logging wrapper handles the same failure. Exercise… |
| Testing (Property / Proof) | ⚠️ Warning | The change introduces invariants over route counts, middleware order, preparation transitions, and repeated connections. `build_route_chains` iterates over every handler and middleware layer, while `W… | Add and recommend a substantive Rust `proptest` or bounded model check. Generate bounded route counts, middleware sequences, and prepared-connection counts. Assert one transform per route and middleware layer, no further transforms after pr… |
| Observability | ⚠️ Warning | The pull request changes operational route-cache behaviour without adding the required production metrics. The diff removes `OnceCell<Arc<HashMap<u32, HandlerService<E>>>>` from `WireframeApp` and add… | Add bounded production observability for the preparation transition. Add counters for preparation outcomes and prepared-connection use, plus a duration histogram for preparation; use stable labels such as `outcome="success"` or `outcome="fa… |
| Performance And Resource Use | ⚠️ Warning | Fail the performance check because the pull request introduces repeated middleware-chain construction on the deprecated `WireframeApp` connection path. The base revision cached `build_chains()` in `On… | Preserve one-time route-chain reuse for the deprecated compatibility path. Restore a compatibility-only cache on `WireframeApp` (or use an equivalent cached prepared representation) and invalidate it when `route` or `wrap` changes builder r… |
| Concurrency And State | ⚠️ Warning | Add an interleaving test for the newly shared `PreparedApp` path. `PreparedApp` is moved into `Arc` and its route services are borrowed by multiple spawned connection tasks (`examples/support/runtime_… | Add a `PreparedApp` integration test that starts at least two connections concurrently with `tokio::join!` or spawned tasks. Use a barrier or controlled async middleware/handler to force overlap. Assert that each connection receives its own… |

**Comment by wafflecat-df12** [comment]
@coderabbitai review

**Comment by pandalump** [comment]
@coderabbitai review

File: docs/developers-guide.md [hunk 10]
```diff
@@ -234,6 +248,23 @@ Install Whitaker through the standalone installer described in the
...
+The `Coverage (main)` workflow in
+`.github/workflows/coverage-main.yml` runs on pushes to `main`. After the test
+suite succeeds, it generates a ratcheted LCOV report and uploads that report to
+CodeScene. The workflow checks out `leynos/wireframe`, so CodeScene records the
+coverage under the repository identity `github.com/leynos/wireframe`, and
+targets project `68308` explicitly.
+
+The upload reads `CS_ACCESS_TOKEN` from the repository secret and passes it to
+the upload action through the job environment; do not put the token in a
+workflow argument or log it. The pull-request workflow's CodeScene coverage
```
[the remark this answers is not in the document]
Comment (leynos) [reply docs/developers-guide.md]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: docs/v0-3-0-to-v0-4-0-migration-guide.md [hunk 11]
```diff
@@ -0,0 +1,221 @@
...
+
+## Migrate byte-handling APIs
+
+The prepared-application transition is independent of the zero-copy byte
+migration. The v0.4 byte-facing APIs use `bytes::Bytes` (or the `PayloadBytes`
+wrapper) for read-only hand-offs and an explicit edit-on-demand operation for
+mutation. The following examples show the required shape of the migration. The
+editor method names are illustrative until roadmap item 12.1.2 finalizes the
+public editing API; the compatibility helper names are defined by
+[ADR 009](adr-009-vec-u8-migration-rollout.md).
```
[the remark this answers is not in the document]
Comment (leynos) [reply docs/v0-3-0-to-v0-4-0-migration-guide.md]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: docs/v0-3-0-to-v0-4-0-migration-guide.md [hunk 12]
```diff
@@ -0,0 +1,221 @@
...
+
+### Custom codecs
+
+Codecs should store payloads as `Bytes` when possible and override
+`frame_payload_bytes` to return a cheap clone. The `wrap_payload` argument is
+already `Bytes`, so only the frame type and extraction methods need changing:
+
+```rust
+// Before: a custom frame owns a Vec<u8> payload.
+struct MyFrame {
```
[the remark this answers is not in the document]
Comment (leynos) [reply docs/v0-3-0-to-v0-4-0-migration-guide.md]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

**Comment by pandalump** [comment]
@coderabbitai review

**Comment by leynos** [comment]
@coderabbitai Have the following now been resolved?

#### wireframe_testing/src/helpers/fragment_drive.rs (1)

`340-364`: _🎯 Functional Correctness_ | _🟡 Minor_ | _⚡ Quick win_
**Preserve connection-processing errors in the chunked driver.**
`drive_chunked_internal` requires a unit-returning server future. Thus `drive_with_partial_fragments` uses `handle_connection`, which logs errors from `handle_connection_result` and returns `()`. Stream-processing and handler errors can be lost. Make the chunked path propagate `io::Result<()>`, update its shared callers, and pass `handle_connection_result`. Retain the existing deprecated-lint expectation for the legacy `WireframeApp` API.

##### 🤖 Detailed instructions

Use a wyvern agent team to verify each finding against the current code and only fix it if needed. Fix any documentation issues using a scribe agent team. Fix only still-valid issues, skip the rest with a brief reason, keep changes minimal, and validate. Use a scrutineer agent for execution of tests and summarizing the results.

In `@wireframe_testing/src/helpers/fragment_drive.rs` around lines 340 - 364, The
chunked driver currently discards connection-processing errors because
drive_chunked_internal accepts a unit-returning future and the caller uses
handle_connection. Update drive_chunked_internal and its shared callers to
propagate io::Result<()> and pass WireframeApp::handle_connection_result
instead, while preserving the existing deprecated-lint expectation for the
legacy WireframeApp API.

_Source: Learnings_

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

Do not treat warnings as optional or aspirational. Where a change is out of scope for this PR, propose a GitHub issue unless one exists already. (Treat o11y, code safety, documentation and validation coverage as in scope).

### ❌ Failed checks (2 errors, 3 warnings)

| Check name | Status | Explanation | Resolution |
| :---: | :--- | :--- | :--- |
| Unit Architecture | ❌ Error | Inject the connection timing dependency before merge. The new `PreparedApp::handle_connection_result` command calls `Instant::now()` directly at `src/app/prepared_app.rs:213` and later calls `started_… | Add a narrow monotonic time-source interface for connection instrumentation (use leynos/monotony for this purpose), with a production implementation backed by `Instant`. Make `handle_connection_result` delegate to a private or crate-visible time-source-aware implementation, and … |
| Rust Compiler Lint Integrity | ❌ Error | The change adds two `#[expect(dead_code)]` suppressions to real `PreparedApp` fields, `app_data` and `push_dlq`, at `src/app/prepared_app.rs:53-75`. The current code only moves these fields during pre… | Keep the fields only if the planned connection-runtime work requires them. Add a source-level tracker reference to each narrow expectation, such as `see issue `[#643](https://github.com/leynos/wireframe/issues/643)`` (and link the roadmap item if appropriate), with a comment that states when… |
| Testing (Unit And Behavioural) | ⚠️ Warning | Add end-to-end coverage for the changed prepared TCP path. The new tests exercise `PreparedApp` through `tokio::io::duplex` and exercise real TCP only through the unchanged `WireframeServer` legacy pa… | Add an integration test that prepares an application, binds a real `TcpListener`, accepts a `TcpStream`, dispatches it through the prepared connection path, and asserts the response and one-time middleware transformation. Exercise the share… |
| Performance And Resource Use | ⚠️ Warning | The change introduces a repeated route-preparation regression in the deprecated `WireframeApp` compatibility API. The base implementation initialized `self.routes` once with `OnceCell` and only cloned… | Restore a reusable compatibility route cache for `WireframeApp::handle_connection_result` and `handle_connection`, or remove those public compatibility paths and migrate all supported callers to one `PreparedApp`. In the cache approach, ini… |
| Concurrency And State | ⚠️ Warning | Fail: preserve connection state cleanup across cancellation and panic paths. `PreparedApp::handle_connection_result` creates setup state and calls `process_connection`; `process_connection` invokes `o… | Add an explicit connection finalization design before exposing the prepared runtime as a task entry point. Pass a cooperative cancellation signal into the connection task and ensure cancellation exits processing before awaiting teardown. Ca… |

**Comment by wafflecat-df12** [comment]
@coderabbitai Thank you for the classification table. Follow-up on each row, against head `34f8d50`.

### Unit Architecture — confirmed still resolved

`PreparedApp::handle_connection_result` (`src/app/prepared_app.rs:211`) delegates to the private
`handle_connection_with_clock` (`src/app/prepared_app.rs:220`), which takes `&dyn monotony::MonotonicClock`.
Production passes `StdMonotonicClock` (`src/app/prepared_app.rs:215`); `src/app/prepared_app_tests.rs`
passes `FixedMonotonicClock`. The only remaining `Instant::now()` is inside
`SystemPreparationTimeSource::start` (`src/app/prepared_app.rs:105`), which is the *preparation*
instrumentation time source — a separate seam from the connection clock, and already injectable through
`prepare_with_time_source`.

### Rust Compiler Lint Integrity — confirmed still resolved

`PreparedApp::app_data` (`src/app/prepared_app.rs:56-61`) and `PreparedApp::push_dlq`
(`src/app/prepared_app.rs:73-78`) both carry narrow `#[expect(dead_code, reason = "...")]` attributes whose
stated reasons reference issue #643, with the surrounding comments linking the same tracker explicitly.

### Testing (Unit and Behavioural) — now resolved

Added `prepared_app_serves_real_tcp_connection_without_retransforming` to `tests/prepared_app.rs`. It
satisfies every point of your recommended prompt:

- Builds a `WireframeApp` with two routes and two transform-counting, tag-emitting middleware layers
  (`A` outer, `B` inner), then calls `prepare().await` exactly once.
- Binds a real `tokio::net::TcpListener` to `127.0.0.1:0` and reads back the ephemeral address.
- Spawns one accept task that accepts a real `TcpStream` and calls `PreparedApp::handle_connection_result`
  on it — the public prepared path, not the legacy `WireframeServer` path.
- Connects a real `tokio::net::TcpStream` client, sends one valid encoded frame, shuts down the write side
  so the server sees EOF, and decodes the response.
- Asserts the response payload is `[b'X', b'A', b'B', b'B', b'A']`, pinning middleware *order* as well as
  presence.
- Asserts the transform counter equals `ROUTES * MIDDLEWARE_LAYERS` after preparation and is **unchanged**
  after the TCP connection is handled — i.e. the connection performs no re-transformation.
- Every wait (accept, connection processing, client connect, response read, task join) is wrapped in a
  bounded `tokio::time::timeout`. The `JoinHandle` result and the connection-processing result are both
  propagated, so a failed join and a failed connection both surface as test failures.

### Performance and Resource Use — out of scope for this PR, tracked by #642

As your table states, `WireframeApp::handle_connection_result` still rebuilds route chains per connection.
That is the documented deprecated compatibility behaviour, and the migration guide records it. This PR
deliberately does **not** add a compatibility cache. Server startup preparation and shared prepared server
roots remain in issue #642.

### Concurrency and State — out of scope for this PR, tracked by #643

This PR does not implement cancellation or panic lifecycle finalization, and I am **not** claiming that
teardown-on-cancellation or teardown-on-panic is implemented. What the PR does verify is teardown after an
ordinary processing error, and reuse of prepared services across overlapping connections. The
`ConnectionRuntime` ownership and finalization contract remains issue #643, and this lifecycle requirement
should not be marked resolved until #643 implements and tests it.

### Validation performed on `34f8d50`

| Command | Result |
|---|---|
| `cargo fmt --all -- --check` | exit 0 |
| `cargo test --test prepared_app prepared_app_serves_real_tcp_connection_without_retransforming` | 1 passed, 0 failed |
| `cargo test --test prepared_app` | 5 passed, 0 failed |
| `cargo clippy -p wireframe --all-targets -- -D warnings` | exit 0 |
| `cargo clippy -p wireframe_testing --all-targets -- -D warnings` | exit 0 |

The new test was also run 25 times consecutively in each of two repeat batches, with 0 failures, to
confirm the real-socket path is deterministic rather than timing-dependent.

No further review request is being made.

**Comment by pandalump** [comment]
@coderabbitai approve
